src/test/java/com/commercetools/sync/products/helpers/VariantReferenceResolverTest.java

```diff
@@ -598,4 +625,159 @@ void resolveReferences_WithNoPriceReferences_ShouldResolveAttributeReferences()
+        // preparation
+        final ProductVariantDraft withNestedTextAttributesOnly = SphereJsonUtils
+            .readObjectFromResource(NESTED_ATTRIBUTE_WITH_SET_OF_PRODUCT_REFERENCE_ATTRIBUTES,
+                ProductVariantDraft.class);
+
+        // test
+        final ProductVariantDraft resolvedAttributeDraft =
+            referenceResolver.resolveReferences(withNestedTextAttributesOnly)
+                             .toCompletableFuture()
+                             .join();
+        // assertions
+        assertThat(resolvedAttributeDraft.getAttributes()).isNotNull();
+
+        final JsonNode value = resolvedAttributeDraft.getAttributes().get(0).getValue();
+        assertThat(value).isInstanceOf(ArrayNode.class);
+        final ArrayNode resolvedNestedAttributes = (ArrayNode) value;
+
+        final Map<String, JsonNode> resolvedNestedAttributesMap = StreamSupport
+            .stream(resolvedNestedAttributes.spliterator(), false)
+            .collect(Collectors.toMap(jsonNode -> jsonNode.get("name").asText(), jsonNode -> jsonNode));
+
+        assertReferenceSetAttributeValue(resolvedNestedAttributesMap,
+            "nested-attribute-1-name", 2, PRODUCT_ID, "product");
+        assertReferenceAttributeValue(resolvedNestedAttributesMap,
+            "nested-attribute-2-name", PRODUCT_ID, "product");
+        assertReferenceAttributeValue(resolvedNestedAttributesMap,
+            "nested-attribute-3-name", PRODUCT_ID, "product");
+    }
+
+    private void assertReferenceAttributeValue(
+        @Nonnull final Map<String, JsonNode> attributeDraftMap,
+        @Nonnull final String attributeName,
+        @Nonnull final String referenceId,
+        @Nonnull final String referenceTypeId) {
+
+        assertThat(attributeDraftMap.get(attributeName)).isNotNull();
+        assertThat(attributeDraftMap.get(attributeName).get("value")).isNotNull();
+        assertThat(attributeDraftMap.get(attributeName).get("value").get("id").asText()).isEqualTo(referenceId);
+        assertThat(attributeDraftMap.get(attributeName).get("value").get("typeId").asText()).isEqualTo(referenceTypeId);
+    }
+
+    private void assertReferenceSetAttributeValue(
+        @Nonnull final Map<String, JsonNode> attributeDraftMap,
+        @Nonnull final String attributeName,
+        final int numberOfReferences,
+        @Nonnull final String referenceId,
+        @Nonnull final String referenceTypeId) {
+
+        assertThat(attributeDraftMap.get(attributeName)).isNotNull();
+        final JsonNode value = attributeDraftMap.get(attributeName).get("value");
+        assertThat(value).isInstanceOf(ArrayNode.class);
+
+        final ArrayNode valueAsArrayNode = (ArrayNode) value;
+        assertThat(valueAsArrayNode).hasSize(numberOfReferences);
+        assertThat(valueAsArrayNode).allSatisfy(jsonNode -> {
+            assertThat(jsonNode.get("id").asText()).isEqualTo(referenceId);
```

Comment:
maybe use VariantReferenceResolver constants for "id" and "typeId" strings.

Reply:
7d8b250

636205c